ledger/appcow.go

```diff
@@ -566,16 +566,21 @@ func applyStorageDelta(data basics.AccountData, aapp storagePtr, store *storageD
 		case deallocAction:
 			delete(owned, aapp.aidx)
 		case allocAction, remainAllocAction:
-			// TODO verify this assertion
 			// note: these should always exist because they were
-			// at least preceded by a call to PutWithCreatable?
+			// at least preceded by a call to PutWithCreatable
 			params, ok := owned[aapp.aidx]
 			if !ok {
 				return basics.AccountData{}, fmt.Errorf("could not find existing params for %v", aapp.aidx)
 			}
 			params = params.Clone()
-			if store.action == allocAction {
-				// TODO does this ever accidentally clobber?
+			if store.action == allocAction ||
+				store.action == remainAllocAction && params.GlobalState == nil {
+				// allocate KeyValue for
+				// 1) app creation and global write in the same app call
+				// 2) global state writing into empty global state
+				params.GlobalState = make(basics.TealKeyValue)
+			}
+			if store.action == remainAllocAction && params.GlobalState == nil {
```

Comment:
in the case of `store.action == remainAllocAction && params.GlobalState == nil` you're going to reallocate it twice. why?

Reply:
in fact, you could have used the predicate `store.action == allocAction || params.GlobalState == nil`, this this switch branch is only for allocAction and remainAllocAction.